### PR TITLE
test: enable hubble tls in ci

### LINKED ISSUE #13093: CI: enable mTLS for Hubble and Hubble Relay in CI tests
PR #12906 and #12900 introduced mTLS support for Hubble server and Hubble Relay. The Hubble Relay CI tests should be updated to default to an mTLS connection between Hubble server instances and Hubble Relay.

### PR DESCRIPTION
**Pull request by kaworu**
Enable TLS for Hubble and verify that the expected certificate is served.

### CODE REVIEW

**Review by rolinh — APPROVED** [review]
Awesome! It's nice that all it takes to enable TLS in the CI is to remove `global.hubble.tls.enabled=false`!

File: test/k8sT/hubble.go [hunk 1]
```diff
@@ -198,6 +198,19 @@ var _ = Describe("K8sHubbleTest", func() {
 			res.ExpectContains(`hubble_flows_processed_total{protocol="TCP",subtype="to-endpoint",type="Trace",verdict="FORWARDED"}`)
 		})
 
+		It("Test TLS certificate", func() {
+			cmd := "cat /var/lib/cilium/tls/hubble/server.crt"
+			res := kubectl.ExecPodCmd(helpers.CiliumNamespace, ciliumPodK8s1, cmd)
+			res.ExpectSuccess("%q cannot read the hubble server TLS certificate", ciliumPodK8s1)
+			expected := string(res.GetStdOut().Bytes())
+
+			cmd = fmt.Sprintf("openssl s_client -connect localhost:4244 -showcerts -servername %s.%s.hubble-grpc.cilium.io | openssl x509 -outform PEM", helpers.K8s1, namespaceForTest)
```
Comment (gandro) [inline test/k8sT/hubble.go]: I think you want to use `default` (as in the "default" cluster name) instead of `namespaceForTest` for the servername here?

**Comment by kAworu** [comment]
test-me-please

**Review by gandro — APPROVED** [review]

File: test/k8sT/hubble.go [hunk 2]
```diff
@@ -198,6 +198,20 @@ var _ = Describe("K8sHubbleTest", func() {
 			res.ExpectContains(`hubble_flows_processed_total{protocol="TCP",subtype="to-endpoint",type="Trace",verdict="FORWARDED"}`)
 		})
 
+		It("Test TLS certificate", func() {
+			cmd := "cat /var/lib/cilium/tls/hubble/server.crt"
```
Comment (christarazi) [inline test/k8sT/hubble.go]: I think we should limit the amount of shelling out that we do in tests, and prefer using native mechanisms, like in this case, using `ioutil.ReadFile`.

File: test/k8sT/hubble.go [hunk 3]
```diff
@@ -198,6 +198,20 @@ var _ = Describe("K8sHubbleTest", func() {
...
+			res := kubectl.ExecPodCmd(helpers.CiliumNamespace, ciliumPodK8s1, cmd)
+			res.ExpectSuccess("%q cannot read the hubble server TLS certificate", ciliumPodK8s1)
+			expected := string(res.GetStdOut().Bytes())
+
+			nodeName, clusterName := helpers.K8s1, "default"
+			cmd = fmt.Sprintf("openssl s_client -connect localhost:4244 -showcerts -servername %s.%s.hubble-grpc.cilium.io | openssl x509 -outform PEM", nodeName, clusterName)
```
Comment (christarazi) [inline test/k8sT/hubble.go]: Is `localhost:4244` static, as in it won't be modified? Would it be worth parameterizing this, or is it safe to assume?

File: test/k8sT/hubble.go [hunk 4]
```diff
@@ -198,6 +198,20 @@ var _ = Describe("K8sHubbleTest", func() {
 			res.ExpectContains(`hubble_flows_processed_total{protocol="TCP",subtype="to-endpoint",type="Trace",verdict="FORWARDED"}`)
 		})
 
+		It("Test TLS certificate", func() {
+			cmd := "cat /var/lib/cilium/tls/hubble/server.crt"
+			res := kubectl.ExecPodCmd(helpers.CiliumNamespace, ciliumPodK8s1, cmd)
+			res.ExpectSuccess("%q cannot read the hubble server TLS certificate", ciliumPodK8s1)
```
Comment (christarazi) [inline test/k8sT/hubble.go]: Nit: These msgs are usually used to grep to the test code, so having a format specifier will throw it off. Can we generalize the msg to say "Cilium pod cannot read ..."?

**Review by christarazi — CHANGES_REQUESTED** [review]
LGTM overall, just a few minor comments.

File: test/k8sT/hubble.go (hunk 2 above)
Comment (kaworu) [reply test/k8sT/hubble.go]: In this case `cmd` is executed into the Cilium pod. I'm unsure how `ioutil.ReadFile` could be used to do that, but I would be happy to add a `ReadFile` in _test/helpers/wrappers.go_ if that is what you meant.

File: test/k8sT/hubble.go (hunk 4 above)
Comment (kaworu) [reply test/k8sT/hubble.go]: Make sense, thanks!

File: test/k8sT/hubble.go (hunk 3 above)
Comment (kaworu) [reply test/k8sT/hubble.go]: I would say it is safe to assume, but we can take a middle ground and use the constant representing the default value.

**Comment by kAworu** [comment]
test-me-please

File: test/k8sT/hubble.go (hunk 2 above)
Comment (christarazi) [reply test/k8sT/hubble.go]: :facepalm: You're right. The thought didn't even cross my mind that this is running inside a cluster. :sweat_smile:

File: test/k8sT/hubble.go [hunk 5]
```diff
@@ -198,6 +199,21 @@ var _ = Describe("K8sHubbleTest", func() {
...
 		})
 
+		It("Test TLS certificate", func() {
+			certpath := "/var/lib/cilium/tls/hubble/server.crt"
+			res := kubectl.ExecPodCmd(helpers.CiliumNamespace, ciliumPodK8s1, helpers.ReadFile(certpath))
+			res.ExpectSuccess("Cilium pod cannot read the hubble server TLS certificate")
+			expected := string(res.GetStdOut().Bytes())
+
+			serverName := fmt.Sprintf("%s.default.hubble-grpc.cilium.io", helpers.K8s1)
+			cmd := helpers.OpenSSLShowCerts("localhost", defaults.ServerPort, serverName)
```
Comment (christarazi) [inline test/k8sT/hubble.go:209]: I think this is a nice change as well!

File: test/helpers/wrappers.go [hunk 6]
```diff
@@ -172,3 +172,20 @@ func PythonBind(addr string, port uint16, proto string) string {
 		`/usr/bin/python3 -c 'import socket; socket.socket(%s).bind((%q, %d))`,
 		strings.Join(opts, ", "), addr, port)
 }
+
+// ReadFile returns the string representing a cat command to read the file at
+// the give path.
+func ReadFile(path string) string {
+	return fmt.Sprintf("cat %q", path)
+}
```
Comment (christarazi) [inline test/helpers/wrappers.go:180]: Thanks for adding the helpers!

**Review by christarazi — APPROVED** [review]
LGTM! Thanks for adding those extra pieces, makes the tests more readable.

**Pull request merged**